File: service/autosportlabs/comms/bluetooth/bluetoothconnection.py
from jnius import autoclass

BluetoothAdapter = autoclass('android.bluetooth.BluetoothAdapter')
UUID = autoclass('java.util.UUID')

class PortNotOpenException(Exception):
    pass

class CommsErrorException(Exception):
    pass

class BluetoothConnection(object):
    ser = None
    recv_stream = None
    send_stream = None
    socket = None
    error_message = None
    
    def __init__(self, **kwargs):
        pass
            
    def get_available_ports(self):
        return ['RaceCapturePro']
                
    def isOpen(self):
        return self.socket != None
    

    def open(self, port):
        error_message = None
        
        try:
            paired_devices = BluetoothAdapter.getDefaultAdapter().getBondedDevices().toArray()
            socket = None
            print('attempting to open port::: ' + port)
            for device in paired_devices:
                if device.getName() == port:
                    print('waiting to connect to device ' + port)
                    socket = device.createRfcommSocketToServiceRecord(UUID.fromString("00001101-0000-1000-8000-00805F9B34FB"))
                    print('after wait')
                    recv_stream = socket.getInputStream()
                    send_stream = socket.getOutputStream()
                    break
            if socket != None:
                print('attempting to connect socket')
                socket.connect()
                print('socket connected')                    
                self.socket = socket
                self.recv_stream = recv_stream
                self.send_stream = send_stream
                error_message = None
            else:
                error_message = 'Could not detect device {}'.format(port)
        except Exception as e:
            print('error opening socket ' + str(e))
            error_message = 'Error opening Bluetooth socket: {}'.format(str(e))

        if error_message != None:
            raise Exception(error_message)
        if self.socket == None:
            raise Exception("Timed out opening Bluetooth port")
                                    
    def close(self):
        try:
            self.socket.close()
        except Exception as e:
            print('Error closing Bluetooth socket: ' + str(e))
        finally:
            self.socket = None
            self.recv_stream = None
            self.send_stream = None


    def read(self, count):
        recv_stream = self.recv_stream
        if recv_stream == None: raise PortNotOpenException()
        c = recv_stream.read()
        if c == -1:
            self.close()
            raise Exception('Socket reported EOF')
        else:
            return chr(c)

    def read_line(self):
        msg = ''
        while True:
            c = self.read(1)
            if c == '':
                return None
            msg += c
            if msg[-2:] == '\r\n':
                msg = msg[:-2]
                return msg
    
    def write(self, data):
        try:
            self.send_stream.write(data)
        except Exception as e:
            print('write error; stream not active ' + str(e))
            raise
    
    def flushInput(self):
        pass
    
    def flushOutput(self):
        self.send_stream.flush()

File: java/src/com/autosportlabs/racecapture/BluetoothConnection.java
package com.autosportlabs.racecapture;
import android.bluetooth.BluetoothAdapter;
import android.bluetooth.BluetoothDevice;
import android.bluetooth.BluetoothSocket;
import android.util.Log;

import java.util.UUID;
import java.util.Set;
import java.util.List;
import java.util.ArrayList;
import java.io.BufferedReader;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.io.OutputStream;
import java.lang.reflect.Method;

/**
 * Bluetooth connection bridge for RaceCature app
 * @author brent
 *
 */
public class BluetoothConnection {
	private static final UUID SerialPortServiceClass_UUID = UUID.fromString("00001101-0000-1000-8000-00805F9B34FB");

	private volatile BluetoothSocket socket;
	private volatile InputStream recv_stream;
	private volatile OutputStream send_stream;
	private volatile BufferedReader reader;
	boolean mAllowInsecureConnections = false;
	Object lock = new Object();
	
	static BluetoothConnection g_instance;
	
	private BluetoothConnection(){
	}
	
	static public BluetoothConnection createInstance(){
		if (g_instance == null){
			g_instance = new BluetoothConnection();
		}
		return g_instance;
	}
	
	private class ConnectThread extends Thread{
		
		BluetoothDevice device = null;
		String errorMessage = null;
		
        public ConnectThread(BluetoothDevice device) {
			this.device = device;
        }
        
        
        public void run(){
        	try{
	        	if ( mAllowInsecureConnections ) {
	        		Log.i("BluetoothConnection", "create socket insecure mode");
	        		Method method;
	        		method = this.device.getClass().getMethod("createRfcommSocket", new Class[] { int.class } );
	        		BluetoothConnection.this.socket = (BluetoothSocket) method.invoke(this.device, 1);  
	        	}
	        	else {
	        		Log.i("BluetoothConnection", "create socket secure mode");
	        		BluetoothConnection.this.socket = this.device.createRfcommSocketToServiceRecord( SerialPortServiceClass_UUID );
	        	}
	            if (BluetoothConnection.this.socket != null){
	            	Log.i("BluetoothConnection", "Got a socket");
	            	BluetoothConnection.this.socket.connect();
	            	Log.i("BluetoothConnection", "Socket setup");                
	            }        	
        	}
        	catch(Exception e){
        		this.errorMessage = e.getMessage();
        	}
        }
	}
	
	/*
	 * Opens the bluetooth connection with the specified port name
	 */
	public void open(String port){
		synchronized(this.lock){
			String error_message = null;
	        try{
	        	BluetoothAdapter adapter = BluetoothAdapter.getDefaultAdapter();
	        	adapter.cancelDiscovery();
	        	Set<BluetoothDevice> paired_devices = adapter.getBondedDevices();
	            for (BluetoothDevice device: paired_devices){
	            	String name = device.getName();
	                if (name.equals(port)){
	                	Log.i("BluetoothConnection", "trying " + name);
	                	
	                    ConnectThread ct = new ConnectThread(device);
	                    ct.start();
	                    ct.join(5000);
	                    error_message = ct.errorMessage;
	                    break;
	                }
	            }
	            if (this.socket != null){
		            this.recv_stream = this.socket.getInputStream();
		            this.send_stream = this.socket.getOutputStream();
		            this.reader = new BufferedReader(new InputStreamReader(this.recv_stream,"UTF-8"));
		            Log.i("BluetoothConnection", "Socket Ready");
	            }
	            else{
	                error_message = "Could not detect device " + port + ": " + String.valueOf(error_message);
	            }            
	        }
	        catch(Exception e){
	            error_message = "Error opening Bluetooth socket: " + e.getMessage();
	        }
	            
	        if (this.socket == null){
	            throw new RuntimeException("Error opening Bluetooth port: " + String.valueOf(error_message));
	        }
		}
	}
	
	/**
	 * Closes the current Bluetooth connection
	 */
	public void close(){
		synchronized(this.lock){		
			try{
	        	Log.i("BluetoothConnection", "Closing Socket");
	        	if (this.socket != null){
	                this.socket.close();        		
	        	}
	        }
	        catch (Throwable e){
	        }
	        finally{
	            this.socket = null;
	            this.recv_stream = null;
	            this.send_stream = null;
	            this.reader = null;
	        }
		}
	}
	
	/**
	 * Returns true if the connection is active
	 */
	public boolean isOpen(){
		return this.socket != null;
	}
	
	/**
	 * reads a line from the socket. If the socket is closed or no data is available, returns null
	 */
	public String readLine(){
		String line = null;
		synchronized(this.lock){
			try{
				if (this.reader != null){
					line =  this.reader.readLine();
				}
			}
			catch(Throwable e){}
		}
		return line;
	}
	
	/**
	 * Writes the specified data to the socket. Returns true if the write succeeded. 
	 */
	public boolean write(String data){
		synchronized(this.lock){
			try{
				if (this.send_stream != null){
					this.send_stream.write(data.getBytes());
					return true;
				}
				else{
					return false;
				}
			}
			catch(Throwable e){
				Log.i("BluetoothConnection", "Error writing data " + data + ": " + e.getMessage());
				this.close();
				return false;
			}
		}
	}

	public String[] getAvailableDevices(){
        BluetoothAdapter bluetoothAdapter = BluetoothAdapter.getDefaultAdapter();
        Set<BluetoothDevice> pairedDevices = bluetoothAdapter.getBondedDevices();

        List<String> devices = new ArrayList<String>();
        for(BluetoothDevice bt : pairedDevices)
            devices.add(bt.getName());

        return devices.toArray(new String[devices.size()]);
	}
}
